fenicsadapter/fenicsadapter.py

```diff
@@ -1,759 +1,458 @@
+        """
+        write_function_type = determine_function_type(write_function)
+        assert (write_function_type in list(FunctionType))
 
-        if self._write_function_type is FunctionType.SCALAR:
-            self._interface.write_block_scalar_data(self._write_data_id, self._vertex_ids, self._write_data)
-        elif self._write_function_type is FunctionType.VECTOR:
-            if self._can_apply_2d_3d_coupling():
-                # in 2d-3d coupling z dimension is set to zero
-                precice_write_data = np.column_stack((self._write_data[:, 0],
-                                                      self._write_data[:, 1],
-                                                      np.zeros(self._n_vertices)))
+        write_data = convert_fenics_to_precice(write_function, self._coupling_mesh_vertices)
 
-                assert(precice_write_data.shape[0] == self._n_vertices and
-                       precice_write_data.shape[1] == self._dimensions)
+        write_data_id = self._interface.get_data_id(self._config.get_write_data_name(),
+                                                    self._interface.get_mesh_id(self._config.get_coupling_mesh_name()))
 
-                self._interface.write_block_vector_data(self._write_data_id, self._vertex_ids, precice_write_data)
-                    
-            elif self._fenics_dimensions == self._dimensions:
-                self._interface.write_block_vector_data(self._write_data_id, self._vertex_ids, self._write_data)
+        if write_function_type is FunctionType.SCALAR:
+            self._interface.write_block_scalar_data(write_data_id, self._vertex_ids, write_data)
+        elif write_function_type is FunctionType.VECTOR:
+            if self._apply_2d_3d_coupling:
+                # in 2d-3d coupling z dimension is set to zero
+                precice_write_data = np.column_stack((write_data[:, 0], write_data[:, 1], np.zeros(self._n_vertices)))
+                assert (precice_write_data.shape[0] == self._n_vertices and
+                        precice_write_data.shape[1] == self._interface.get_dimensions())
+                self._interface.write_block_vector_data(write_data_id, self._vertex_ids, precice_write_data)
+            elif self._fenics_dimensions == self._interface.get_dimensions():
+                self._interface.write_block_vector_data(write_data_id, self._vertex_ids, write_data)
             else:
                 raise Exception("Dimensions don't match.")
         else:
             raise Exception("Rank of function space is neither 0 nor 1")
 
-    def _read_block_data(self):
-        """ Reads data from preCICE. Depending on the dimensions of the simulation (2D-3D Coupling, 2D-2D coupling or
-        Scalar/Vector write function) read_data is converted.
-
-        Note: For quasi 2D fenics in a 3D coupled simulation the y component of the vectors is deleted.
+    def initialize(self, coupling_subdomain, mesh, read_function, function_space, dimensions=2):
```

Comment:
```suggestion
    def initialize(self, coupling_subdomain, mesh, function_space, read_function=None, dimensions=2):
```
We should at make the `read_function` an optional argument, since one might use explicit coupling, where the FEniCS participant only writes data. Then the user still has to call `initialize`, but cannot provide a `read_function`.

Further below in `initialize` we should ensure that if no `read_function` is provided that actually the user wants to do explicit coupling and did not just forget it. 

For now, if would just raise an exception:
```
if not read_function:
    raise Exception("Please provide a read_function. Explicit coupling is not yet supported for the FEniCS adapter.")
```
We can implement and test explicit coupling at a later point in time, but we should make sure that our interface to the user already supports it.

Reply:
Note: I addad an issue on this: https://github.com/precice/fenics-adapter/issues/72

Reply:
Added to current design in the form of one Exception:
```
raise Exception("Please provide a read_function. Explicit coupling is not yet supported for the FEniCS adapter.")
```